Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -4168,6 +4168,40 @@ void Document::make_unsalvageable([[maybe_unused]] String reason)
     set_salvageable(false);
 }
 
+struct DocumentDestructionState : public GC::Cell {
+    GC_CELL(DestructionState, GC::Cell);
+
+    DocumentDestructionState(GC::Ref<Document> document, size_t remaining, GC::Ptr<GC::Function<void()>> after)
+        : remaining_children(remaining)
+        , document(document)
+        , after_all(after)
+    {
+    }
+
+    virtual void visit_edges(GC::Cell::Visitor& visitor) override
+    {
+        Base::visit_edges(visitor);
+        visitor.visit(document);
+        visitor.visit(after_all);
+    }
+
+    void increment_destroyed()
+    {
+        --remaining_children;
+        if (remaining_children > 0)
+            return;
```

Comment:
In this new case, instead of hanging we never destroy the document, right? Maybe we could add a timeout as well to detect this happening?

Reply:
yeah, having timeout is a good idea. I added one that logs FIXME in case 15s was not enough to destroy all documents. but just to clarify: `increment_destroyed()` is supposed to be invoked multiply times until `remaining_children` is 0 and that's what happens now on https://www.rottentomatoes.com/ that was previously stuck in spin until.